File: frontend/src/services/LeaderboardService.js
import api from '../api/axiosConfig';

const LeaderboardService = {
  getLeaderboard: async (category, sortBy) => {
    try {
      const params = {};
      if (category) params.category = category;
      if (sortBy) params.sortBy = sortBy;
      const response = await api.get('/v1/leaderboard', { params });
      return response.data;
    } catch (error) {
      console.error('Error fetching leaderboard:', error);
      throw error;
    }
  },

  getUserStats: async () => {
    try {
      const response = await api.get('/v1/leaderboard/user-stats');
      return response.data;
    } catch (error) {
      console.error('Error fetching user leaderboard stats:', error);
      throw error;
    }
  },

  getWeeklyLeaderboard: async (weekStart, weekEnd, category, sortBy) => {
    try {
      const params = {};
      if (weekStart && weekEnd) {
        params.weekStart = weekStart;
        params.weekEnd = weekEnd;
      }
      if (category) params.category = category;
      if (sortBy) params.sortBy = sortBy;
      const response = await api.get('/v1/leaderboard/weekly', { params });
      return response.data;
    } catch (error) {
      console.error('Error fetching weekly leaderboard:', error);
      throw error;
    }
  },

  getWeeklyLeaderboardHistory: async (weekStart, weekEnd) => {
    try {
      const response = await api.get('/v1/leaderboard/weekly/history', {
        params: { weekStart, weekEnd }
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching weekly leaderboard history:', error);
      throw error;
    }
  },

  getMonthlyLeaderboard: async (monthStart, monthEnd, category, sortBy) => {
    try {
      const params = {};
      if (monthStart && monthEnd) {
        params.monthStart = monthStart;
        params.monthEnd = monthEnd;
      }
      if (category) params.category = category;
      if (sortBy) params.sortBy = sortBy;
      const response = await api.get('/v1/leaderboard/monthly', { params });
      return response.data;
    } catch (error) {
      console.error('Error fetching monthly leaderboard:', error);
      throw error;
    }
  },

  getMonthlyLeaderboardHistory: async (monthStart, monthEnd) => {
    try {
      const response = await api.get('/v1/leaderboard/monthly/history', {
        params: { monthStart, monthEnd }
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching monthly leaderboard history:', error);
      throw error;
    }
  },

  getYearlyLeaderboard: async (year, category, sortBy) => {
    try {
      const params = {};
      if (year) params.year = year;
      if (category) params.category = category;
      if (sortBy) params.sortBy = sortBy;
      const response = await api.get('/v1/leaderboard/yearly', { params });
      return response.data;
    } catch (error) {
      console.error('Error fetching yearly leaderboard:', error);
      throw error;
    }
  },

  getYearlyLeaderboardHistory: async (year) => {
    try {
      const response = await api.get('/v1/leaderboard/yearly/history', {
        params: { year }
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching yearly leaderboard history:', error);
      throw error;
    }
  }
};

export default LeaderboardService;

File: backend/src/main/java/com/carbonfootprint/service/LeaderboardService.java
package com.carbonfootprint.service;

import com.carbonfootprint.dto.leaderboard.LeaderboardResponseDto;
import com.carbonfootprint.dto.leaderboard.WeeklyLeaderboardResponseDto;

import java.time.LocalDate;

public interface LeaderboardService {
    LeaderboardResponseDto getLeaderboard(String currentUserEmail, String category, String sortBy);
    
    WeeklyLeaderboardResponseDto getWeeklyLeaderboard(String currentUserEmail, LocalDate weekStart, LocalDate weekEnd, String category, String sortBy);
    
    java.util.List<com.carbonfootprint.dto.leaderboard.WeeklyLeaderboardEntryDto> getWeeklyLeaderboardHistory(LocalDate weekStart, LocalDate weekEnd);
    
    com.carbonfootprint.dto.leaderboard.MonthlyLeaderboardResponseDto getMonthlyLeaderboard(String currentUserEmail, LocalDate monthStart, LocalDate monthEnd, String category, String sortBy);
    
    java.util.List<com.carbonfootprint.dto.leaderboard.MonthlyLeaderboardEntryDto> getMonthlyLeaderboardHistory(LocalDate monthStart, LocalDate monthEnd);
    
    com.carbonfootprint.dto.leaderboard.YearlyLeaderboardResponseDto getYearlyLeaderboard(String currentUserEmail, int year, String category, String sortBy);
    
    java.util.List<com.carbonfootprint.dto.leaderboard.YearlyLeaderboardEntryDto> getYearlyLeaderboardHistory(int year);
    
    // Optional: a method to recalculate or clear cache if needed manually
    void evictLeaderboardCache();
    
    com.carbonfootprint.dto.leaderboard.UserLeaderboardStatsDto getUserLeaderboardStats(String currentUserEmail);
}
